src/pid/environ.rs

```diff
@@ -0,0 +1,78 @@
+//! Process initial environment from `/proc/[pid]/environ`.
+
+use std::ffi::{OsStr, OsString};
+use std::fs::File;
+use std::io::{Result, Read};
+use std::os::unix::ffi::OsStrExt;
+use std::path::Path;
+
+use parsers::map_result;
+use nom::IResult;
+
+use libc::pid_t;
+
+/// A list of environment variables and their values.
+pub type Environ = Vec<(OsString, OsString)>;
+
+/// Parses the provided buffer.
+fn parse_environ(input: &[u8]) -> IResult<&[u8], Environ> {
+    // Parse 'key=value" pair.
+    named!(
+        pair<&[u8], (&[u8], &[u8])>,
+        tuple!(take_until_and_consume!("="), take_until!("\0"))
+    );
+    // Parse (key=value\0) list.
+    named!(
+        multi<&[u8],Vec<(&[u8], &[u8])> >,
+        terminated!(
+            separated_list!(tag!("\0"), pair),
+            tag!("\0")
+        )
+    );
+    multi(input).map(|r| {
+        r.into_iter()
+            .map(|(x, y)| {
```

Comment:
If you move this map call into the `pair` function, the number of vectors constructed can be reduced from two to one.

Reply:
The full code context has disappeared (thanks GitHub), but I think the idea was that you could use a `map!` call inside the parsing code, instead of the `Iterator::map` call here.